netann/chan_status_manager.go

```diff
@@ -0,0 +1,595 @@
+		// populate the entry in the chanStates map.
+		curState, err := m.getOrInitChanStatus(c.FundingOutpoint)
+		if err != nil {
+			log.Errorf("Unable to retrieve chan status for "+
+				"Channel(%v): %v", c.FundingOutpoint, err)
+			continue
+		}
+
+		// If the channel's status is not ChanStatusEnabled, we are
+		// done.  Either it is already disabled, or it has been marked
+		// ChanStatusPendingDisable meaning that we have already
+		// scheduled the time at which it will be disabled.
+		if curState.Status != ChanStatusEnabled {
+			continue
+		}
+
+		// If our bookkeeping shows the channel as active, sample the
+		// htlcswitch to see if it believes the link is also active. If
+		// so, we will skip marking it as ChanStatusPendingDisabled.
+		chanID := lnwire.NewChanIDFromOutPoint(&c.FundingOutpoint)
+		if m.cfg.IsChannelActive(chanID) {
+			continue
+		}
+
+		// Otherwise, we discovered that this link was inactive within
+		// the switch. Compute the time at which we will send out a
+		// disable if the peer is unable to reestablish a stable
+		// connection.
+		disableTime := time.Now().Add(m.cfg.ChanDisableTimeout)
+
+		log.Debugf("Marking channel(%v) pending-inactive",
+			c.FundingOutpoint)
+
+		m.chanStates.markPendingDisabled(c.FundingOutpoint, disableTime)
+	}
+}
+
+// disableInactiveChannels scans through the set of monitored channels, and
+// broadcast a disable update for any pending inactive channels whose
+// SendDisableTime has been superseded by the current time.
+func (m *ChanStatusManager) disableInactiveChannels() {
+	// Now, disable any channels whose inactive chan timeout has elapsed.
+	now := time.Now()
+	for outpoint, state := range m.chanStates {
+		// Ignore statuses that are not in the pending-inactive state.
+		if state.Status != ChanStatusPendingDisabled {
+			continue
+		}
+
+		// Ignore statuses for which the disable timeout has not
+		// expired.
+		if state.SendDisableTime.After(now) {
+			continue
+		}
+
+		log.Infof("Announcing channel(%v) disabled "+
+			"[detected]", outpoint)
+
+		// Sign an update disabling the channel.
+		err := m.signAndSendNextUpdate(outpoint, true)
+		if err != nil {
+			log.Errorf("Unable to sign update disabling "+
+				"channel(%v): %v", outpoint, err)
+			continue
+		}
+
+		// Record that the channel has now been disabled.
+		m.chanStates.markDisabled(outpoint)
+	}
+}
+
+// fetchChannels returns the working set of channels managed by the
+// ChanStatusManager. The returned channels are filtered to only contain public
+// channels.
+func (m *ChanStatusManager) fetchChannels() ([]*channeldb.OpenChannel, error) {
+	allChannels, err := m.cfg.DB.FetchAllOpenChannels()
+	if err != nil {
+		return nil, err
+	}
+
+	// Filter out private channels.
+	var channels []*channeldb.OpenChannel
+	for _, c := range allChannels {
+		// We'll skip any private channels, as they aren't used for
+		// routing within the network by other nodes.
+		if c.ChannelFlags&lnwire.FFAnnounceChannel == 0 {
```

Comment:
Not a blocker, but this could have been hidden behind this new interface. With that path, then the `ChanStatusManager` doesn't need to know how to interpret the bits of the channel flag, or what an advertised channel even is.